crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -2924,6 +2924,23 @@ impl<'db> TypeInferenceBuilder<'db> {
             | Type::TypeVar(..)
             | Type::AlwaysTruthy
             | Type::AlwaysFalsy => {
+                if let Type::NominalInstance(instance) = object_ty {
+                    let dataclass_params = match instance.class() {
+                        ClassType::NonGeneric(cls) => cls.dataclass_params(self.db()),
+                        ClassType::Generic(_) => None,
+                    };
+                    let frozen = dataclass_params
+                        .is_some_and(|params| params.contains(DataclassParams::FROZEN));
+                    if frozen && emit_diagnostics {
+                        if let Some(builder) = self.context.report_lint(&INVALID_ASSIGNMENT, target)
+                        {
+                            builder.into_diagnostic(format_args!(
+                                "Property `{attribute}` defined in `{ty}` is read-only",
```

Comment:
It looks like at the moment we'll emit this for any attempt to set any attribute on a frozen dataclass. If the attribute does not exist on the dataclass, this seems like a confusing error message, because the attribute is not in fact defined on the type at all. Would it be better in that case to not emit this message, and just the unresolved-attribute one below?

Reply:
@carljm ah thanks - this makes sense. i also checked how mypy handles this for `INVALID_ATTRIBUTE_ACCESS` above that and appears to be the same - they'll only emit the read-only diagnostic for frozen dataclasses if it looks like it would be assignable otherwise:

```python
from dataclasses import dataclass
from typing import ClassVar

@dataclass(frozen=True)
class Starship:
    damage: ClassVar[int] = 1

s = Starship()
s.damage = 2 # error: Cannot assign to class variable "damage" via instance  [misc]
```

changed the logic so we only emit the read-only diagnostic if it looks like the attribute assignment would otherwise be valid